src/utils/sbml_utils.cpp

```diff
@@ -0,0 +1,132 @@
+/******************************************************************************
+ *                                                                            *
+ *    Copyright 2020   Lawrence Livermore National Security, LLC and other    *
+ *    Whole Cell Simulator Project Developers. See the top-level COPYRIGHT    *
+ *    file for details.                                                       *
+ *                                                                            *
+ *    SPDX-License-Identifier: MIT                                            *
+ *                                                                            *
+ ******************************************************************************/
+
+#include "utils/sbml_utils.hpp"
+
+#if defined(WCS_HAS_CONFIG)
+#include "wcs_config.hpp"
+#else
+#error "no config"
+#endif
+
+
+#if defined(WCS_HAS_SBML)
+#include <sbml/SBMLTypes.h>
+#include <sbml/common/extern.h>
+#endif // defined(WCS_HAS_SBML)
+
+#include <iostream> 
+#include <string>
+#include <unordered_set>
+
+namespace wcs {
+/** \addtogroup wcs_utils
+ *  *  @{ */
+
+sbml_utils::sbml_utils()
+: m_type(_undefined_)
+{
+}
+
+#if defined(WCS_HAS_SBML)
+
+std::unordered_set<std::string> sbml_utils::find_undeclared_reactant_species(const libsbml::Model& model, const libsbml::Reaction& reaction)
+{
+  using model_parameters = std::unordered_set<std::string>;
+  typename model_parameters::const_iterator pit;
+  model_parameters pset;
+
+  using reaction_reactants= std::unordered_set<std::string>;
+  typename reaction_reactants::const_iterator rit;
+  reaction_reactants rset;
+
+  reaction_reactants undeclared_reactants;
+  
+  using model_compartments= std::unordered_set<std::string>;
+  typename model_compartments::const_iterator cit;
+  model_compartments cset;
+
+  using reaction_symbols= std::unordered_set<std::string>;
+  ///typename reaction_symbols::const_iterator sit;
+  reaction_symbols symbol_set;
+  
+
+  const libsbml::ListOfSpecies* specieslist = model.getListOfSpecies();
+  
+  // create an unordered_set for model parameters
+  const libsbml::ListOfParameters* parameterslist = model.getListOfParameters();
```

Comment:
check if the parameterlist is nullptr.
However, it is probably ok to omit checking parameterlist->get(pi) because pi is less than parameterlist->size()
use LIBSBML_CPP_NAMESPACE:: instead of libsbml::